skipchain/skipchain.go

```diff
@@ -664,16 +888,104 @@ func (s *Service) newBlockEnd(sb *SkipBlock) bool {
 	return true
 }
 
+// authenticate searches if this node or any follower-node can verify the
+// schnorr-signature.
+func (s *Service) authenticate(msg []byte, sig []byte) bool {
+	if err := schnorr.Verify(Suite, s.ServerIdentity().Public, msg, sig); err == nil {
+		return true
+	}
+	s.storageMutex.Lock()
+	defer s.storageMutex.Unlock()
+	for _, sb := range s.Storage.Follow {
+		for _, si := range sb.Roster.List {
+			if err := schnorr.Verify(Suite, si.Public, msg, sig); err == nil {
+				return true
+			}
+		}
+	}
+	for _, cl := range s.Storage.Clients {
+		if err := schnorr.Verify(Suite, cl, msg, sig); err == nil {
+			return true
+		}
+	}
+	return false
+}
+
+// blockIsFriendly searches if all members of the new block are followed
+// by this node.
+func (s *Service) blockIsFriendly(sb *SkipBlock) bool {
+	if AuthSkipchain == 0 {
+		return true
+	}
+	// accept all blocks that are already stored with us.
+	if s.getByID(sb.SkipChainID()) != nil {
+		return true
+	}
+	s.storageMutex.Lock()
+	defer s.storageMutex.Unlock()
+	// Also accept blocks that are stored in the FollowIDs
+	for _, id := range s.Storage.FollowIDs {
+		if id.Equal(sb.SkipChainID()) {
+			return true
+		}
+	}
+	switch AuthSkipchain {
+	case 1:
```

Comment:
It seems like case 2 is missing here? (Not implemented yet?)

Reply:
Implemented differently now.